### PR TITLE
Bringing structure to the Lcalc AST

### PR DESCRIPTION
**Pull request by denismerigoux**
After #235 and in preparation for #204, this PR brings the scope let-binding structure and metadata from Lcalc down to Dcalc.

## Changelog

* [x] Introduced constructors for boxed terms in Dcalc
* [x] Fixed doc generation 
* [x] Improves Clerk such that it recognizes some Make flags  
* [x] Dcalc refactoring 
* [x] Refactor the Lcalc AST
* [x] Adapt translation from Dcalc to Lcalc

### CODE REVIEW

File: build_system/clerk_driver.ml [hunk 1]
```diff
@@ -700,16 +709,37 @@ let add_test_builds
...
+  match makeflags with
+  | None -> ""
+  | Some makeflags ->
+      let ignore_rex = Re.Pcre.regexp "\\bi\\b" in
+      let has_ignore = Re.Pcre.pmatch ~rex:ignore_rex makeflags in
+      let jobs_rex = Re.Pcre.regexp "-j(\\d)" in
+      let number_of_jobs =
+        try
+          int_of_string
+            (Re.Pcre.get_substring (Re.Pcre.exec ~rex:jobs_rex makeflags) 1)
```
Comment (AltGr) [inline build_system/clerk_driver.ml:722]: Nitpicking here, but why use `Pcre` ?
It's generally considered much more legible to use the building functions of `Re` (although the regexps here remain pretty trivial).

```suggestion
      let ignore_rex = Re.(compile @@ word (char 'i')) in
      let has_ignore = Re.execp ignore_rex makeflags in
      let jobs_rex = Re.(compile @@ seq [str "-j"; group (rep digit)]) in
      let number_of_jobs =
        try
          int_of_string
            (Re.get (Re.exec jobs_rex makeflags) 1)
```

**Review by AltGr — APPROVED** [review]
All good, well structured and I didn't find any mistake. Two very minor comments/questions, not really related to the core of the PR.

File: build_system/clerk_driver.ml [hunk 2]
```diff
@@ -749,7 +779,7 @@ let driver
             (Format.formatter_of_out_channel out)
             (add_root_test_build ninja ctx.all_file_names ctx.all_test_builds);
           close_out out;
-          let ninja_cmd = "ninja test -f " ^ ninja_output in
+          let ninja_cmd = "ninja " ^ ninja_flags ^ " test -f " ^ ninja_output in
           Cli.debug_print "executing '%s'..." ninja_cmd;
           Sys.command ninja_cmd
```
[the remark this answers is not in the document]
Comment (denismerigoux) [reply build_system/clerk_driver.ml:784]: Good remark, there are other instances of `Sys.command` across the codebase. Does `Unix.execvp` work on Windows if we ever manage to compile Catala to a Windows executable ?

[the remark this answers is not in the document]
Comment (AltGr) [reply build_system/clerk_driver.ml:782]: `fork` doesn't, so it's better to rely on `Unix.open_process*` (and restrict uses of `waitpid`) for windows compat. One-way `exec` is ok though.

**Pull request merged**